src/prog_algs/state_estimators/kalman_filter.py

```diff
@@ -66,7 +65,19 @@ def __init__(self, model, x0, measurement_eqn = None, **kwargs):
 
         self.filter = kalman.KalmanFilter(num_states, num_measurements, num_inputs)
 
-        self.filter.x = np.array([[x0[key]] for key in model.states])
+        parameter_Q_bool = 'Q' not in self.parameters # cache check
+        if isinstance(x0, dict):
+            warnings.warn(f"Warning: Use UncertainData type if estimating filtering with uncertain data.")
+            self.filter.x = np.array([[x0[key]] for key in model.states])
+            if parameter_Q_bool:
+                self.parameters['Q'] = np.diag([1.0e-3 for i in x0.keys()])
+        elif isinstance(x0, UncertainData):
+            self.filter.x = np.array([[x0.mean()[key]] for key in model.states])
+            if parameter_Q_bool:
+                self.parameters['Q'] = x0.cov() # need something to start off, use covariance 2d numpy array
+        else:
+            raise TypeError("TypeError: x0 initial state must be of type {{dict, UncertainData}}")
+
         self.filter.P = self.parameters['Q'] / 10
```

Comment:
```suggestion
```
Remove this, instead assign P directly above (only using Q when x0 is a dict)

Reply:
Fixed in Teams discussion